api/src/routes/certificate.test.ts

```diff
@@ -0,0 +1,307 @@
+import { Certification } from '../../../shared/config/certification-settings';
+import {
+  defaultUserEmail,
+  defaultUserId,
+  devLogin,
+  setupServer,
+  superRequest
+} from '../../jest.utils';
+
+describe('certificate routes', () => {
+  setupServer();
+  describe('Authenticated user', () => {
+    let setCookies: string[];
+
+    // Authenticate user
+    beforeAll(async () => {
+      setCookies = await devLogin();
+    });
+
+    describe('PUT /certificate/verify', () => {
+      beforeEach(async () => {
+        await fastifyTestInstance.prisma.user.updateMany({
+          where: { email: defaultUserEmail },
+          data: {
+            completedChallenges: [],
+            name: 'fcc',
+            isRespWebDesignCert: false,
+            isJsAlgoDataStructCert: false,
+            isFrontEndLibsCert: false,
+            is2018DataVisCert: false,
+            isRelationalDatabaseCertV8: false,
+            isApisMicroservicesCert: false,
+            isQaCertV7: false,
+            isSciCompPyCertV7: false,
+            isDataAnalysisPyCertV7: false,
+            isInfosecCertV7: false,
+            isMachineLearningPyCertV7: false,
+            isCollegeAlgebraPyCertV8: false,
+            isFoundationalCSharpCertV8: false,
+            username: 'fcc'
+          }
+        });
+      });
+
+      test('should return 500 if no certSlug', async () => {
+        const response = await superRequest('/certificate/verify', {
+          method: 'PUT',
+          setCookies
+        }).send({ certSlug: undefined });
+        expect(JSON.parse(response.text)).toMatchObject({
+          code: 'FST_ERR_FAILED_ERROR_SERIALIZATION'
+        });
```

Comment:
Couple of things.

`response.body` should give you the parsed body, so you don't need `JSON.parse` here. There's a few other cases later on where you can simplify this.

Also, we want to send an error response with `flash.wrong-name`, right? Regardless, a serialization error should never occur, since it means our handler has tried to return something unexpected

Reply:
This is `response`:

```bash
{"header": {"access-control-allow-credentials": "true", "access-control-allow-headers": "Origin, X-Requested-With, Content-Type, Accept, Csrf-Token", "access-control-allow-methods": "GET, PUT, POST, DELETE", "access-control-allow-origin": "https://www.freecodecamp.org", "cache-control": "no-store", "connection": "close", "content-length": "643", "content-security-policy": "frame-ancestors 'none'", "content-type": "application/json; charset=utf-8", "date": "Thu, 12 Oct 2023 11:50:17 GMT", "set-cookie": ["csrf_token=PR8jxEZp-eEgwvv3I8KblTOhZzLBBApKvhv1FMA8RE6UmGUJ6d0Y; Domain=localhost; Path=/; SameSite=Strict"], "vary": "Origin, Accept-Encoding", "x-content-type-options": "nosniff", "x-frame-options": "DENY"}, "req": {"data": {"certSlug": "responsive-web-design"}, "headers": {"content-type": "application/json", "cookie": ["_csrf=n5tHCOG-NNVDfVlbuWgF-B44; Path=/; HttpOnly; SameSite=Strict", "csrf_token=4ND3axq1-SCLBQoGxcfDIXfqNo0NqQEKdvEE99F-nSI6vexDdjM8; Domain=localhost; Path=/; SameSite=Strict", "sessionId=jdIUR6yfm4hrrPk_2rAM-A_7mnQ0i5zB.BDi1lbSZGPdHXVayT21SL6qOXDwLdHPcscETkbFxCFg; Path=/; Expires=Thu, 12 Oct 2023 12:50:17 GMT; HttpOnly"], "csrf-token": "4ND3axq1-SCLBQoGxcfDIXfqNo0NqQEKdvEE99F-nSI6vexDdjM8", "origin": "https://www.freecodecamp.org"}, "method": "PUT", "url": "http://127.0.0.1:44511/certificate/verify"}, "status": 400, "text": "{\"response\":{\"type\":\"info\",\"message\":\"flash.incomplete-steps\",\"variables\":{\"name\":\"Responsive Web Design\"}},\"isCertMap\":{\"isRespWebDesignCert\":false,\"isJsAlgoDataStructCert\":false,\"isFrontEndLibsCert\":false,\"is2018DataVisCert\":false,\"isApisMicroservicesCert\":false,\"isInfosecQaCert\":false,\"isQaCertV7\":false,\"isInfosecCertV7\":false,\"isFrontEndCert\":false,\"isBackEndCert\":false,\"isDataVisCert\":false,\"isFullStackCert\":false,\"isSciCompPyCertV7\":false,\"isDataAnalysisPyCertV7\":false,\"isMachineLearningPyCertV7\":false,\"isRelationalDatabaseCertV8\":false,\"isCollegeAlgebraPyCertV8\":false,\"isFoundationalCSharpCertV8\":false},\"completedChallenges\":[]}"}
```

There is no `response.body`.

Reply:
EDIT (_without editing_): Turns out there is a `body`, it is just a getter.